File: backend/email_scraper.py
"""Email extraction from business websites"""

import re
import time

import requests


def scrape_emails_from_websites(results):
    """Visit each website and extract contact email using regex.
    Updates results in-place with 'email' field."""
    for i, p in enumerate(results):
        website = p.get("website", "")
        if not website:
            p["email"] = ""
            continue

        try:
            print(f"  [{i + 1}/{len(results)}] Scraping: {website}")
            res = requests.get(
                website,
                timeout=8,
                headers={
                    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
                },
            )
            match = re.search(
                r"[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+", res.text
            )
            p["email"] = match.group(0).lower() if match else ""
            if p["email"]:
                print(f"    📧 Found: {p['email']}")
            else:
                print(f"    ❌ No email found")
        except Exception as e:
            p["email"] = ""
            print(f"    ⚠️ Error: {str(e)[:80]}")

        time.sleep(0.5)  # Be respectful to websites

    return results

File: lib/places/email-scraper.ts
// Email extraction from business websites (ported from backend/email_scraper.py)

import type { PlaceResult } from "./types"

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms))
const EMAIL_RE = /[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+/

export async function scrapeEmailsFromWebsites(
  results: PlaceResult[]
): Promise<PlaceResult[]> {
  for (const p of results) {
    const website = p.website ?? ""
    if (!website) {
      p.email = ""
      continue
    }

    try {
      const controller = new AbortController()
      const timeout = setTimeout(() => controller.abort(), 8000)
      const res = await fetch(website, {
        headers: {
          "User-Agent":
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
        },
        signal: controller.signal,
      })
      clearTimeout(timeout)
      const text = await res.text()
      const match = text.match(EMAIL_RE)
      p.email = match ? match[0].toLowerCase() : ""
    } catch (e) {
      p.email = ""
      console.log(`[v0] Email scrape error for ${website}: ${String(e).slice(0, 80)}`)
    }

    await sleep(500)
  }

  return results
}
